### PR TITLE
Implement bech32 p2wpkh funding addresses

### LINKED ISSUE #648: Feature Request: bech32 funding address
`cli/lightning-cli newaddr` currently generates a legacy p2sh address. It would be nice and slightly cheaper if it could also generate a `bech32` funding address.

### PR DESCRIPTION
**Pull request by jb55**
* Add optional addresstype param to newaddr, which can be one of:

    - bech32
    - p2sh-segwit
    - nothing (defaults to p2sh-segwit)

The naming here mirrors bitcoind

* txfilter already looks for p2wpkh outputs, so we're covered there

* Tests require a version of bitcoind with a segwit wallet

Fixes #648

### CODE REVIEW

File: wallet/walletrpc.c [hunk 1]
```diff
@@ -316,10 +316,28 @@ static void json_newaddr(struct command *cmd,
...
+		return;
+	}
+
+	if (!addrtype || json_tok_streq(buffer, addrtype, "p2sh-segwit"))
+		is_p2wpkh = false;
+	else if (json_tok_streq(buffer, addrtype, "bech32"))
+		is_p2wpkh = true;
+	else {
+		command_fail(cmd,
+			     "Invalid address type param (expected segwit/p2wpkh or p2sh)");
```
Comment (ZmnSCPxj) [inline wallet/walletrpc.c]: Error message is misleading --- it seems from the error message that you accept "segwit/p2wpkh" or "p2sh", but really you accept instead "p2sh-segwit" or "bech32").

File: wallet/walletrpc.c [hunk 2]
```diff
@@ -341,22 +359,36 @@ static void json_newaddr(struct command *cmd,
 
 	txfilter_add_derkey(cmd->ld->owned_txfilter, ext.pub_key);
 
-	redeemscript = bitcoin_redeem_p2sh_p2wpkh(cmd, &pubkey);
-	sha256(&h, redeemscript, tal_count(redeemscript));
-	ripemd160(&p2sh, h.u.u8, sizeof(h));
+	if (is_p2wpkh) {
+		/* segwit_addr_encode's internal buffer size is 65, so just use that */
+		out = tal_arr(cmd, char, 65);
```
Comment (ZmnSCPxj) [inline wallet/walletrpc.c]: Better to follow the header comments of segwit_addr_encode: https://github.com/ElementsProject/lightning/blob/master/common/bech32.h#L34

    out = tal_arr(cmd, char, 73 + strlen(get_chainparams(cmd->id)->bip173_name);

File: wallet/walletrpc.c (hunk 1 above)
Comment (jb55) [reply wallet/walletrpc.c]: Good catch!

File: wallet/walletrpc.c (hunk 2 above)
Comment (jb55) [reply wallet/walletrpc.c]: 👍

**Comment by jb55** [comment]
I'll just rebase with the tests in this one

**Comment by Sjors** [comment]
I was able to fund the LN wallet using `lightning-cli newaddr bech32` and able to open a channel.

Probably unrelated, but might be worth testing anyway: I'm not able to pay any invoice ("Could not find a route") and it doesn't seem to be broadcasting any channel funding transactions beyond the first one (even `getrawtransaction` doesn't see them).

Maybe update the documentation to specify that for testing the bitcoin core version must be 0.16 or master?

**Comment by Ulmo** [comment]
Does the red x mean this wasn't integrated?  I'd love this option.

**Comment by cdecker** [comment]
Seems the travis-ci builder doesn't have an up to date `bitcoind`, I'll upgrade to 0.15.1 tomorrow.

**Comment by cdecker** [comment]
Upgraded `bitcoind` in the builder CI image to 0.15.1, but still fails. Are bech32 addresses supported only in 0.16.0? If that's the case I vote to postpone rolling out this PR until 0.16.0 gets released, I don't want to test against an unreleased version.

**Comment by ZmnSCPxj** [comment]
https://bitcointalk.org/index.php?topic=2488576.0

Not yet in 0.15.1. Concur that it is somewhat difficult if we need an unreleased version for testing. Frustrating though as we already have a bech32 parser.

Or maybe.... we can withdraw to a bech32 addr already. So how about:

1. Have bitcoind send to a P2SH-P2WPKH address we own.
2. Do `lightning-cli withdraw all $ADDR` to an address we control.
3. Check it appears in listfunds under the P2WPKH bech33 addr.

**Comment by cdecker** [comment]
Very good idea @ZmnSCPxj, I tried implementing that but it appears c-lightning can't read regtest addresses for some reason:

```
RPC call failed: Could not parse destination address
```

With the generated address being `bcrt1q2awwrr0kvs4n7zg7e9mpwhm9t9f2hhr6glvc8k`.

**Comment by ZmnSCPxj** [comment]
Ah no. At the time I was implementing bech32 withdraw, we were using `tb1` only, not `bcrt1`. Should be possible to add though.

**Comment by jb55** [comment]
I'm ok with waiting for a 0.16 release

**Comment by Sjors** [comment]
Needs rebase (especially with the libwally changes)

**Comment by rustyrussell** [comment]
Rebased, let's see how CI does...

**Comment by rustyrussell** [comment]
Python bindings broke, not sure best way to fix it.  @cdecker ?

**Comment by cdecker** [comment]
That's really a weird issue, it seems that for some reason we are pulling in some form of cache from #884 which then results in us using the RPC code with explicit arguments (!?!?!). Testing by creating a new branch on my repo reports the correct error of bech32 not being parseable from bitcoind < 0.16.0. Trying to fix around that.

**Comment by jb55** [comment]
Looks like 0.16.0rc2 is out https://bitcoin.org/bin/bitcoin-core-0.16.0/test.rc2/ 

I'll update this PR with the new version for CI

**Comment by cdecker** [comment]
It's probably better not to require developers to run a release candidate just to be able to run tests. I'd therefore prefer the withdraw-to-self approach proposed by @rustyrussell. I also noticed that this needs a rebase, despite GH claiming it's good to go, the JSON API has changed a bit.

**Comment by cdecker** [comment]
Notice that withdraw doesn't understand bech32 addresses yet either, so that test won't work :frowning_face:

**Comment by rustyrussell** [comment]
After quick discussion with @cdecker we decided to include, with test disabled for now....

**Comment by ZmnSCPxj** [comment]
@cdecker withdraw does understand bech32. It just understands bc1 and tb1 addresses, not bcrt1

**Comment by rustyrussell** [comment]
Good point @ZmnSCPxj !  Want to open an issue? :)

**Pull request merged**